src-tauri/src/main.rs

```diff
@@ -72,80 +78,91 @@ async fn control_grabbing(
     frontend_bool: bool,
 ) -> Result<(), ()> {
     state.set_is_listening(frontend_bool);
-    // set_macro_listening(state, frontend_bool);
+    Ok(())
+}
+
+/// Enables or disables the automatic startup of Wootomation at system start.
+fn init_autostart(app_name: &str, set_autolaunch: bool) -> Result<(), Error> {
+    let current_exe = std::env::current_exe().context("current EXE not found")?;
+
+    let auto_start = auto_launch::AutoLaunchBuilder::new()
+        .set_app_name(app_name)
+        .set_app_path(current_exe.as_path().to_str().unwrap())
+        .set_use_launch_agent(true)
+        .build()
+        .context("App name is empty, or unsupported OS is used.")?;
+
+    match set_autolaunch {
+        true => auto_start.enable()?,
+        false => {
+            if let Err(e) = auto_start.disable() {
+                match e {
+                    auto_launch::Error::Io(err) => match err.kind() {
+                        std::io::ErrorKind::NotFound => {
+                            trace!("Autostart is already removed, finished checking.");
+                        }
+                        _ => return Err(err.into()),
+                    },
+                    _ => return Err(e.into()),
+                }
+            }
+        }
+    }
     Ok(())
 }
 
 #[tokio::main(flavor = "multi_thread", worker_threads = 10)]
 /// Spawn the backend thread.
 /// Note: this doesn't work on macOS since we cannot give the thread the proper permissions
 /// (will crash on key grab/listen)
-async fn main() {
+async fn main() -> Result<(), Error> {
+    // The optionENV only takes a string literal and can't be forced to take a const of any type.
     let log_level: log::LevelFilter = option_env!("MACRO_LOG_LEVEL")
         .and_then(|s| log::LevelFilter::from_str(s).ok())
         .unwrap_or(log::LevelFilter::Info);
 
-    wooting_macro_backend::MacroBackend::generate_directories();
+    MacroBackend::generate_directories().context("unable to generate config directories")?;
 
     let backend = MacroBackend::default();
 
-    info!("Running the macro backend");
-
     #[cfg(any(target_os = "windows", target_os = "linux"))]
-    backend.init().await;
+    // The backend is run only on Windows and Linux, on macOS it won't work.
+    info!("Running the macro backend");
+    if let Err(e) = backend.init().await {
+        eprintln!("Initialization error: {}", e);
```

Comment:
I wonder if the `error!` should be used here? Surely that would put it in the stderr as well, and would also go through the logging system used by everything else

Reply:
Although, I guess there's a bit of a weirdness here with logging, cus it gets set-up by Tauri, so before that runs, none of the special log handles will actually work?

Reply:
Tweaked to use error! instead of eprintln, seems to be fine?